lib/logger/lib/logger.ex

```diff
@@ -19,6 +19,27 @@ defmodule Logger do
     * Wraps OTP's `error_logger` to prevent it from
       overflowing.
 
+  A common use case is tracking when an important function is called.
+
+      def solve(problem) do
+        Logger.info fn ->
+          "solve/1 called with problem #{inspect problem}"
+        end
+        solve(problem, [])
+      end
+
+  The `Logger.info/2` macro emits the provided message at the `:info`
+  level. There are additional macros for other levels. Refer to the `:level` and
+  `:compile_time_purge_level` configuration options for more details on how
+  levels work.
+
+  Notice the argument passed to `Logger.info` in the above example is a zero
+  argument function. Although the `Logger` macros accept strings which aren't
+  wrapped in functions, it's not recommended in this case. Calling `"... #{inspect
+  problem}"` without wrapping it in a zero argument function will hurt
+  performance, even when then `:info` level is blocked by the configured runtime
+  logging level.
```

Comment:
Actually, I want to elaborate on the first sentence more.

Evaluating a zero argument function is a constant-time operation when the configured `:level` causes the message to be ignored. However, an expensive logging expression that isn't wrapped in a function will be evaluated at runtime either way (unless the whole `Logger` call is compiled out using `:complie_time_level_purge`). Thus, performance can degrade for expensive logging calls that aren't even used.